staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/schema/cel/validation.go

```diff
@@ -251,16 +254,89 @@ func (s *Validator) validateExpressions(ctx context.Context, fldPath *field.Path
 			continue
 		}
 		if evalResult != types.True {
-			if len(rule.Message) != 0 {
-				errs = append(errs, field.Invalid(fldPath, sts.Type, rule.Message))
+			if compiled.MessageExpression != nil {
+				messageExpression, newRemainingBudget, msgErr := evalMessageExpression(ctx, compiled.MessageExpression, rule.MessageExpression, activation, remainingBudget)
+				if msgErr != nil {
+					if msgErr.errType == runtimeCostErr {
+						errs = append(errs, field.InternalError(fldPath, msgErr.err))
+						continue
+					} else {
+						klog.Warning(msgErr.err)
+						errs = append(errs, field.Invalid(fldPath, sts.Type, ruleMessageOrDefault(rule)))
+					}
+				} else {
+					errs = append(errs, field.Invalid(fldPath, sts.Type, messageExpression))
+					remainingBudget = newRemainingBudget
+				}
 			} else {
-				errs = append(errs, field.Invalid(fldPath, sts.Type, fmt.Sprintf("failed rule: %s", ruleErrorString(rule))))
+				errs = append(errs, field.Invalid(fldPath, sts.Type, ruleMessageOrDefault(rule)))
 			}
 		}
 	}
 	return errs, remainingBudget
 }
 
+type _errType int
+
+const (
+	runtimeCostErr _errType = iota
+	evalErr
+)
+
+type messageExpressionErr struct {
+	errType _errType
+	err     error
+}
+
+// evalMessageExpression takes a context, Program and source code representing the MessageExpression, activation,
+// and remaining, runtime cost budget, returns the evaluated message and the remaining budget, or an error if one
+// occurred during evaluation.
+func evalMessageExpression(ctx context.Context, expr celgo.Program, exprSrc string, activation interpreter.Activation, remainingBudget int64) (string, int64, *messageExpressionErr) {
+	evalResult, evalDetails, err := expr.ContextEval(ctx, activation)
+	if evalDetails == nil {
+		return "", -1, &messageExpressionErr{
+			runtimeCostErr,
+			fmt.Errorf("runtime cost could not be calculated for message expression, falling back: %q", exprSrc),
+		}
+	}
+	rtCost := evalDetails.ActualCost()
+	if rtCost == nil {
+		return "", -1, &messageExpressionErr{
+			runtimeCostErr,
+			fmt.Errorf("runtime cost could not be calculated for message expression, falling back: %q", exprSrc),
+		}
+	} else {
+		if *rtCost > math.MaxInt64 || int64(*rtCost) > remainingBudget {
+			return "", -1, &messageExpressionErr{
+				runtimeCostErr,
+				fmt.Errorf("messageExpression evaluation failed due to running out of cost budget, falling back: %q", exprSrc),
+			}
+		}
+	}
+	if err != nil {
+		return "", -1, &messageExpressionErr{
+			evalErr,
+			fmt.Errorf("error evaluating message expression, falling back: %v", err.Error()),
+		}
+	}
+	messageStr, ok := evalResult.Value().(string)
+	if !ok {
+		return "", -1, &messageExpressionErr{
+			evalErr,
+			errors.New("could not convert message expression result to string, falling back"),
+		}
+	}
+	return messageStr, remainingBudget - int64(*rtCost), nil
+}
+
+func ruleMessageOrDefault(rule apiextensions.ValidationRule) string {
```

Comment:
Why not directly using `ruleErrorString`?

Reply:
`ruleMessageOrDefault` preserves the current behavior of printing `failed rule: ...` if we don't have `Message`. There are two cases where we want to have that logic:

*  `MessageExpression` results in an error besides exhausting the runtime cost budget
*  `MessageExpression` is not included at all (the user didn't specify one, etc.)